### ISSUE TITLE
HFSAttributeBTree::getattr crashes

### ISSUE THREAD

**Issue by tomkoen** [darlinghq/darling-dmg#68]
```cpp
leafNode = findLeafNode((Key*) &key, cnidAttrComparator);
```

if findLeafNode returns empty HFSBTreeNode then copy constructor will crash in initFromBuffer. I'd changing  the method code to
```cpp
void initFromBuffer()
{
   if (m_descriptorData.size())
   {
      m_descriptor = reinterpret_cast<BTNodeDescriptor*>(&m_descriptorData[0]);
      m_firstRecordOffset = reinterpret_cast<uint16_t*>(descPtr() + m_nodeSize - sizeof(uint16_t));
   }
}
```

**Comment by jief666** [comment]
I've just tried and it doesn't crash. `&m_descriptorData[0]` is strictly equivalent to `m_descriptorData` and is NULL, so `m_descriptor` == NULL, hence  isInvalid() is true.
`m_firstRecordOffset` == -1

It's ok to cast a pointer equal to NULL. Any `getRecordKey()` on that node will end up badly, of course...

**Comment by tomkoen** [comment]
Without the check m_descriptorData[0] is trying to get the first element of the array when the array is smpty.

**Comment by jief666** [comment]
Yes, I wrote earlier too quick. Seems that vector index out of bounds is undefined.
Clang seems to return NULL, which, by coincidence, works well here.

Must be corrected, yes.

Maybe 
```
	HFSBTreeNode& operator=(const HFSBTreeNode& that)
	{
                if (!that.isInvalid()) {
		    m_descriptorData = that.m_descriptorData;
		    m_nodeSize = that.m_nodeSize;
		    initFromBuffer();
		    return *this;
                }
	}

```

**Comment by tomkoen** [comment]
Thanks for confirming. Quick code to reproduce:

```cpp
HFSBTreeNode retEmptyNode() { return HFSBTreeNode(); }

HFSBTreeNode crashNode;
crashNode = retEmptyNode(); // oops
```

**Comment by jief666** [comment]
Which platform/compiler do you use ?

**Comment by tomkoen** [comment]
Windows, Visual Studio 2013

**Comment by jief666** [comment]
I prefer things not to silently fail. Allowing an index out of bounds, at least in debug mode, is annoying.
At least VC crash !

**Comment by jief666** [comment]
HI @LubosD and @bugaevc.

Should we do a pull request for this, so it's quicker for you ?
I'm trained with pull requests, so I can do that :-)

**Comment by bugaevc** [comment]
Yes please :)

**Comment by jief666** [comment]
Which fix do you prefer ? In `void initFromBuffer()` like tomkoen proposed, or in `HFSBTreeNode& operator=(const HFSBTreeNode& that)`

**Comment by jief666** [comment]
The one in `initBuffer()` is better, don't you think.

**Comment by jief666** [comment]
Done : #70
